src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,104 @@
 package com.es.phoneshop.model.product;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
 import java.util.List;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private long maxId;
+    private List<Product> products;
+    private ReadWriteLock lock;
+
+    public ArrayListProductDao() {
+        lock = new ReentrantReadWriteLock();
+        products = new ArrayList<>();
+        saveSampleProducts();
+    }
+
     @Override
     public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+        lock.readLock().lock();
+        Product result = null;
+        try {
+            result = products.stream()
+                    .filter(product -> id.equals(product.getId()))
+                    .findAny()
+                    .orElseThrow(ProductNotFoundException::new);
+        } catch (ProductNotFoundException e) {
+            e.printStackTrace();
+        } finally {
+            lock.readLock().unlock();
+        }
+        return result;
     }
 
     @Override
     public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+        lock.readLock().lock();
+        List<Product> result;
+        try {
+            result = products.stream()
+                    .filter(product -> product.getPrice() != null)
```

Comment:
Why not to make single filter with 2 conditions? It will increase performance a bit in large streams and also such approach is more popular (up to you to change or not).